File: desktop/gatepath/diag/https_only_probe.py
"""Detects captive setups that pass cleartext HTTP but kill HTTPS.

Re-runs the HTTP verdict via `ProbeContext.active_probe`, and only when HTTP
says *validated* does it try the https-scheme variant of
`ProbeContext.probe_url` via `ProbeContext.http_fetch`. HTTPS failing while
HTTP works (TLS interception, RST-on-443) is the `HttpsOnlyCaptive`
signature.

While the network is still captive (HTTP -> portal) or broken (HTTP ->
error), HTTPS failing tells us nothing new — those cases are Healthy here
and owned by other probes.

Mirror of Android `HttpsOnlyProbe.kt`. Desktop's `ProbeResult.status` is the
string "validated" | "portal" | "error", not a sealed type.
"""

from __future__ import annotations

from gatepath.diag.probe import ProbeContext
from gatepath.diag.report import DiagnosticReport, Healthy, HttpsOnlyCaptive


class HttpsOnlyProbe:
    name = "https_only"

    def run(self, ctx: ProbeContext) -> DiagnosticReport:
        http_result = ctx.active_probe()
        if http_result.status != "validated":
            return Healthy()
        https_url = ctx.probe_url.replace("http://", "https://", 1)
        https_result = ctx.http_fetch(https_url, None)
        if https_result.error is not None:
            return HttpsOnlyCaptive(https_error_message=https_result.error)
        return Healthy()

File: android/app/src/main/java/com/ventouxlabs/gatepath/diag/HttpsOnlyProbe.kt
package com.ventouxlabs.gatepath.diag

import com.ventouxlabs.gatepath.network.ProbeResult

/**
 * Detects captive setups that pass cleartext HTTP but kill HTTPS (TLS
 * interception, RST-on-443): re-runs the HTTP verdict via
 * [ProbeContext.activeProbe], and only when HTTP says *validated* does it try
 * the https-scheme variant of [ProbeContext.probeUrl] via
 * [ProbeContext.httpFetch]. HTTPS failing while HTTP works is the
 * [DiagnosticReport.HttpsOnlyCaptive] signature.
 *
 * While the network is still captive (HTTP → Portal) or broken (HTTP →
 * Error), HTTPS failing tells us nothing new — those cases are Healthy here
 * and owned by other probes. This is the "Phase 4 fan-out" HttpProbe's doc
 * deferred.
 */
class HttpsOnlyProbe : DiagnosticProbe {
    override val name = "https_only"

    override suspend fun run(ctx: ProbeContext): DiagnosticReport {
        when (ctx.activeProbe()) {
            is ProbeResult.Portal, is ProbeResult.Error -> return DiagnosticReport.Healthy
            is ProbeResult.Validated -> Unit
        }
        val httpsUrl = ctx.probeUrl.replaceFirst("http://", "https://")
        val https = ctx.httpFetch(httpsUrl, null)
        return when {
            https.error != null -> DiagnosticReport.HttpsOnlyCaptive(httpsErrorMessage = https.error)
            else -> DiagnosticReport.Healthy
        }
    }
}
